storage/limiter.go

```diff
@@ -0,0 +1,173 @@
+// Copyright JAMF Software, LLC
+
+package storage
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/benbjohnson/clock"
+	"github.com/jamf/regatta/storage/kv"
+)
+
+type burst struct {
+	Tokens   uint64        `json:"a"`
+	Interval time.Duration `json:"i"`
+	ValidTo  time.Time     `json:"v"`
+}
+
+type state struct {
+	Tokens    uint64        `json:"a"`
+	Interval  time.Duration `json:"i"`
+	Remaining uint64        `json:"r"`
+	LastTake  time.Time     `json:"l"`
+	Burst     burst         `json:"b,omitempty"`
+	Ver       uint64        `json:"-"`
+}
+
+type KVLimiter struct {
+	rs    *kv.RaftStore
+	clock clock.Clock
+	mtx   sync.Mutex
+}
+
+func (k *KVLimiter) prefixKey(key string) string {
+	return fmt.Sprintf("ratelimit/%s", key)
+}
+
+func (k *KVLimiter) getState(key string) (state, error) {
+	get, err := k.rs.Get(k.prefixKey(key))
+	s := state{}
+	if err != nil {
+		if errors.Is(err, kv.ErrNotExist) {
+			return s, nil
+		}
+		return s, err
+	}
+	if err := json.Unmarshal([]byte(get.Value), &s); err != nil {
+		return s, err
+	}
+	s.Ver = get.Ver
+	return s, nil
+}
+
+func (k *KVLimiter) setState(key string, new state, version uint64) (state, error) {
+	b, err := json.Marshal(new)
+	if err != nil {
+		return new, err
+	}
+	p, err := k.rs.Set(k.prefixKey(key), string(b), version)
+	if err != nil {
+		return new, err
+	}
+	new.Ver = p.Ver
+	return new, nil
+}
+
+func (k *KVLimiter) WaitFor(ctx context.Context, key string) error {
+	for {
+		_, next, ok, _ := k.Take(key)
+		if ok {
+			return nil
+		}
+		t := time.NewTimer(next.Sub(k.clock.Now()))
+		select {
+		case <-ctx.Done():
+			if !t.Stop() {
+				<-t.C
+			}
+			return ctx.Err()
+		case <-t.C:
+			continue
+		}
+	}
```

Comment:
just thinking out loud - have you considered using a shared instance of Ticker and resetting it with each iteration? Or do you think it is not worth the lower readability due to low "demands" on creation of new instances of the Timer with each iteration?
E.g. something like this:
```suggestion
    t := time.NewTicker(time.Second)
    defer t.Stop()
	for {
		_, next, ok, _ := k.Take(key)
		if ok {
			return nil
		}
		
		t.Reset(next.Sub(k.clock.Now()))    <---- this assumes >0 duration, if 0 can happen, it would need to be checked, omitted for brevity
		
		select {
		case <-ctx.Done():
			return ctx.Err()
		case <-t.C:
			continue
		}
	}
```

Reply:
Given it in most cases tick just once I think this should be OK. The problem with Reset is that it should be called again on either drained chan T or stopped timer and that is becoming too complex for what its worth IMO.